Allow to disable the Crossfader

gtrentalancia opened the pull request:
The Crossfader is not always used, also on most controllers (such as the
Hercules P32 DJ) its knob does not lock in the middle position to avoid
unwanted changes.

As a result it is possible that the Crossfader knob is moved involuntarily
away from its middle position during normal controller operation, causing
unwanted (deck) volume imbalances.

This patch adds an option (checkbox) to the Mixxx Graphical User Interface
so that it is possible to disable the Crossfader functionality (controller
mapping).

When enabling the AutoDJ functionality make sure to also re-enable the
Crossfader if it has been previously disabled (thanks to Daniel Schürmann
for raising this issue).

The translation for the label of the new GUI checkbox is provided for the
Italian language only (and English defaults).

The logic for setting the various other Crossfader options is also fixed
and made consistent to avoid bugs.

---
 res/translations/mixxx_en.ts                   |   19 +++---
 res/translations/mixxx_en_GB.ts                |   19 +++---
 res/translations/mixxx_it.ts                   |   21 ++++--
 src/engine/enginemaster.cpp                    |   10 +++
 src/engine/enginemaster.h                      |    1
 src/library/autodj/autodjprocessor.cpp         |   14 ++++
 src/library/autodj/autodjprocessor.h           |    5 +
 src/preferences/dialog/dlgprefcrossfader.cpp   |   77 ++++++++++++++++++-------
 src/preferences/dialog/dlgprefcrossfader.h     |    4 +
 src/preferences/dialog/dlgprefcrossfaderdlg.ui |    9 ++
 10 files changed, 136 insertions(+), 43 deletions(-)

daschuer commented:
Thank you for this PR and contributing to Mixxx. 

I am unsure about this, because this disables also the AutoDJ feature, which might be surprising for some users. This feature may introduce more headaches than it eliminates. On the other hand I fully understand your issue. 

We could change the implementation that only the controller mapped crossfader is effected. 
This can be done by connect to the valueChangeRequest() signal and reject all signals from the controller engine. The AutoDJ can uses setAndConfirm() to bypass the check. 

If we step back a bit more, we have related issues with the xfader and Auto DJ 
* Allow manual fade if Auto DJ tries to fade as well  https://bugs.launchpad.net/mixxx/+bug/995982
* Special AudoDJ crossfader settings https://bugs.launchpad.net/mixxx/+bug/1576530
* AutoDJ should always use 'mixing' crossfader setting https://bugs.launchpad.net/mixxx/+bug/1662525
* AutoDJ fade out rather than crossfade https://bugs.launchpad.net/mixxx/+bug/1766164

All this can be solved by a kind of xFader controller, that received the input from the controllers xfader and from the Auto DJ via two separate controls. It decides what to do with these input signal and selects instantly the desired xfader curve. It may also handle the case when the xfader orientation of the AutoDJ decks are not correct. 

What do the others think about this?

gtrentalancia commented:
Hello Daniel,

thanks for your feedback!

I am a new Mixxx user therefore I don't know anything about the AutoDJ feature that you mentioned.

I have tested this pull request with the basic mixing functionality and it works fine!

I can look at the code tomorrow and hopefully the pull request can be extended so that the AutoDJ functionality is not affected by the changes that I am proposing!

I am sure it can be done, but it will take me a little bit more to modify the patch according to your advice...

I will get back to you shortly.

Regards,

Guido

gtrentalancia commented:
Provided that I have not managed to look at the code yet, as I am away from my laptop, I thought that if, as you say, the AutoDJ functionality is simlply "disabled" by the changes that I am proposing here, it might just be a very simple matter of giving precedence to the AutoDJ setting over the Crossfader-disabled setting.

In other words, I suppose it is just a matter of changing (pseudocode):

IF (REQUIRE_CROSSFADER_DISABLED) THEN DO_DISABLE_CROSSFADER()

as follows:

IF (!AUTODJ_IS_ENABLED && REQUIRE_CROSSFADER_DISABLED) THEN DO_DISABLE_CROSSFADER()

What do you say?

Guido

ywwg commented:
We recently had a bug where we discussed these issues: https://bugs.launchpad.net/mixxx/+bug/1819758

There are enough ways of disabling the crossfader already that we decided an explicit feature was unnecessary.  One change that might help with is showing the [L, M, R] crossfader position settings for each channel on two-deck mode.  (This is the way users disable the crossfader on hardware mixers, by setting all channels to M or sometimes it's called "thru".)   Currently on LateNight, for example, those settings only appear in 4-deck mode.

daschuer commented:
Thank you for the link to the bug. I didn't found it. Now I have set it to "won't fix" for a better finding.

Since we have here a second voice asking for this feature, there seems to be a a valid demand for it. In combination with the other open bugs, maybe we can refactor the revamp the xFader code carefully.

gtrentalancia commented:
The AutoDJ issue pointed out by Daniel Schürmann (daschuer) has been fixed here:

https://github.com/mixxxdj/mixxx/pull/2071/commits/5fa516ceeecb8c2035e35fcbe04e3d9fa5b5be66

gtrentalancia commented:
Can you please test and then merge ?

uklotzde commented:
Please revert all changed .ts files. I suppose these modifications are neither needed nor desired.

gtrentalancia commented:
Hello Uwe,

thanks for your feedback ! I have now reverted the translations for the new GUI checkbox, as you suggested.

It was meant to be a placeholder to help translators, I wasn't sure about the current development policy...

Regards,

Guido

ywwg commented:
> Can you please test and then merge ?

We are all still discussing how we would like to approach solving this problem. I'm not in favor of the approach in this PR, of creating a new setting.  The workaround you've created to re-enable the setting when AutoDJ is on is confusing, and all the skins will need to be changed to indicate this brand new enabled/disabled state or it would be very confusing.

This is a lot of new code for something that we can already do in Mixxx.  I think all we need to do is make it easier to set channels to M mode.

uklotzde commented:
@gtrentalancia By *revert* I actually meant editing your previous commits by interactively rebasing this branch, followed by a *force push*. Otherwise those intermediate commits might still cause merge conflicts once the generated .ts files are pulled from Transifex.

gtrentalancia commented:
> @gtrentalancia By _revert_ I actually meant editing your previous commits by interactively rebasing this branch, followed by a _force push_. Otherwise those intermediate commits might still cause merge conflicts once the generated .ts files are pulled from Transifex.

uklotzde commented:
I agree that a visible but otherwise non-functional crossfader causes usability issues.

gtrentalancia commented:
@uklotzde I cannot revert and force push now. I could close this pull request and open a new one, but I don't know if it is worth it, because Owen Williams does not seem happy with merging this pull request...

What should I do ? I am happy with the changes that I have developed, the suit my needs, I suppose they suit the needs of several more users if not all users.

But my time is limited, so if it is a matter of just creating a clean new pull request without the translations commits, then let's go for it quickly because I need to move on other projects, otherwise let's leave it and I will use it locally on my own laptop.

Guido

gtrentalancia commented:
> I agree that a visible but otherwise non-functional crossfader causes usability issues.

The same thing happens on stand-alone mixers that feature a crossfader-disable button. So it is just the right solution, exactly what happens in the real world.

gtrentalancia commented:
I don't gain anything if it is merged or not, since I have already developed it for my own use, which is what I need.

But feel free to use it, if you like it !

And, of course, merging it is just a matter of getting rid of the translations commit, then just let me know and I can do that...

Guido

gtrentalancia committed:
The Crossfader is not always used, also on most controllers (such as the
Hercules P32 DJ) its knob does not lock in the middle position to avoid
unwanted changes.

As a result it is possible that the Crossfader knob is moved involuntarily
away from its middle position during normal controller operation, causing
unwanted (deck) volume imbalances.

This patch adds an option (checkbox) to the Mixxx Graphical User Interface
so that it is possible to disable the Crossfader functionality (controller
mapping).

When enabling the AutoDJ functionality make sure to also re-enable the
Crossfader if it has been previously disabled (thanks to Daniel Schürmann
for raising this issue).

The translation for the label of the new GUI checkbox is provided for the
Italian language only (and English defaults).

The logic for setting the various other Crossfader options is also fixed
and made consistent to avoid bugs.

Signed-off-by: Guido Trentalancia <guido@trentalancia.com>
---
 res/translations/mixxx_en.ts                   |   19 +++---
 res/translations/mixxx_en_GB.ts                |   19 +++---
 res/translations/mixxx_it.ts                   |   21 ++++--
 src/engine/enginemaster.cpp                    |   10 +++
 src/engine/enginemaster.h                      |    1
 src/library/autodj/autodjprocessor.cpp         |   14 ++++
 src/library/autodj/autodjprocessor.h           |    5 +
 src/preferences/dialog/dlgprefcrossfader.cpp   |   77 ++++++++++++++++++-------
 src/preferences/dialog/dlgprefcrossfader.h     |    4 +
 src/preferences/dialog/dlgprefcrossfaderdlg.ui |    9 ++
 10 files changed, 136 insertions(+), 43 deletions(-)

gtrentalancia commented:
@uklotzde Uwe, I have reverted the translation commits completely. I hope this helps. Feel free to use it or merge it, as you like !

Regards,

Guido

daschuer commented:
I still see a chance to get this feature in. 
Unfortunately this can't be done in a rush.
Since it is a corner case we have to make sure that it does not disturb the normal users. First we need an agreement how to do it. Hopfully you will have still time to make it real finally:

Here some thoughts:
Mixxx currently suffers the issue that the xFader curve is also used for auto Dj transitions. I see also different use cases and preferences for the GUI xFader.
If you consider for example a steep scratching curve it makes probably only sense with a controller. Disabling one of these xFader is IMHO only very special xFader curve. 

How about this: 
* Use the existing xFader control for controller
* Add a new engine xFader or right/left bus volume that reflects the value used in the engine.
* Add a new xfader control for skins
* Add a new xfader control for GUI
* All three xfaders shall be synced in the first version

This architecture finally allows (in a later step) to introduce individual xFader curves  per auto Dj transition. 
It allows also to take control with the controllers during a transition, or other open bugs.

What do you think?

ywwg commented:
> What do you think?

This still feels like a lot of new code.  Here's another idea:  instead of the AutoDJ controlling the crossfader, it controls the channel faders?  It puts the crossfader in the center and then raises and lowers the individual channel gains.  That way the autodj is totally independent of the crossfader.

gtrentalancia commented:
@daschuer Daniel, what you propose seems an overcomplex to me. At the moment, it works and I am perfectly happy with it, why complicating things and wasting time ? I can't see the benefit. It doesn't interfere with the AutoDJ functionality anymore, I have fixed that ! It is easy, compact, fit for the purpose and manageable...

gtrentalancia commented:
@ywwg Owen, the AutoDJ issue raised by Daniel has been fixed several hours ago here: https://github.com/mixxxdj/mixxx/commit/5fa516ceeecb8c2035e35fcbe04e3d9fa5b5be66

I cannot see any benefit in moving the AutoDJ functionality from Crossfader to the Channel Faders.

It just doesn't interfere anymore with the AutoDJ functionality because when the AutoDJ functionality is enabled, it also re-enable the Crossfader and then when the AutoDJ functionality is eventually turned off, it restores the previous Crossfader state.

I hope it helps.

Regards,

Guido

ywwg commented:
> it also re-enables the Crossfader and then when the AutoDJ functionality is eventually turned off, it restores the previous Crossfader state.

if the user exits the program while auto dj is running, and then restarts it, the setting will not be restored.

> I cannot see any benefit in moving the AutoDJ functionality from Crossfader to the Channel Faders at this point.

The benefit is that the auto dj functionality will be divorced from the crossfader settings so that autodj will always behave predictably. Right now autodj assumes that the crossfader is set up a certain way, and any change the user makes could mess up auto dj. For instance if the crossfader curve is set very sharp, the fade won't be smooth.  And I don't know what happens if the crossfader has been adjusted to inverted "hamster mode" for scratching. The channel faders don't have any of these complex settings so it's a better choice. 

(In general, auto DJ has a lot of legacy problems and needs a lot of work.  This would be a great place to start to really improve mixxx for a lot of users!)

Because you have limited time to work on this change then you are asking us to maintain your new feature forever. Because the core development team must live with the code, we have every right to make sure that it will work well for us in the years to come. In the 10 years I've worked on Mixxx I have gained a very good sense of how much every additional feature and option adds to the complexity of maintaining a stable program.  Daniel has also put in many years and also understands what it means to add and maintain more options.

It is already possible to disable the crossfader as I have explained, so we need a really good reason to add a new option that does the same thing.

gtrentalancia commented:
> > it also re-enables the Crossfader and then when the AutoDJ functionality is eventually turned off, it restores the previous Crossfader state.
> 
> if the user exits the program while auto dj is running, and then restarts it, the setting will not be restored.

When AutoDJ re-enables the Crossfader, it does not save the configuration, therefore if the user quits the program while AutoDJ is enabled, the next time he or she restarts Mixxx, the Crossfader setting is restored from the configuration saved on disk.

I hope this helps.

daschuer commented:
> Here's another idea: instead of the AutoDJ controlling the crossfader, it controls the channel faders? It puts the crossfader in the center and then raises and lowers the individual channel gains. That way the autodj is totally independent of the crossfader.

That is a valid option, I think It depends on the user which AutoDJ style is preferred. Finally it just moved the problem of Auto vs. Manual to a different set of controls.

gtrentalancia commented:
> > I cannot see any benefit in moving the AutoDJ functionality from Crossfader to the Channel Faders at this point.
> 
> The benefit is that the auto dj functionality will be divorced from the crossfader settings so that autodj will always behave predictably.

As far as I know, as long as AutoDJ was always behaving predictably before I introduced this change, it will keep always working predictably, because the patch that I proposed is very simple and does not affect AutoDJ functionality.

> Right now autodj assumes that the crossfader is set up a certain way, and any change the user makes could mess up auto dj. For instance if the crossfader curve is set very sharp, the fade won't be smooth. 

This change does not affect the Crossfader settings either. It just introduces a new setting that allows the user to disable the Crossfader, but the settings are left untouched, therefore the Crossfader behaviour is the same as before with only the added benefit that it can now be disabled.

> And I don't know what happens if the crossfader has been adjusted to inverted "hamster mode" for scratching. The channel faders don't have any of these complex settings so it's a better choice.

As already explained, the Crossfader settings are not changed by this patch. If it was working fine before this change, it will keep working fine after this change !

[...]
 
> It is already possible to disable the crossfader as I have explained, so we need a really good reason to add a new option that does the same thing.

At the moment, it is extermely difficult and not intuititve to understand how to disable the Crossfader without this change !

I am strongly recommend that you adopt it as soon as possible, it has only benefits and no side-effects. Just test it and you will be able to verify it by yourself !

gtrentalancia commented:
The only thing I feel like promising is that, if this Pull Request is merged, I will update the Mixxx Manual accordingly, so that the new feature is properly documented.

daschuer commented:
I have checked this PR and it works as described. However it does not fully resolve the original issue of accidentally touching the hardware xFader.

I think we all agree, that there is an underlying issue, the demand of independent Auto DJ gains.

What could be a solution that is future prove and picks up the work here and fits to @gtrentalancia schedule?

We currently already have the issue that if the user messes with the xFader curve or the deck orientation that it breaks the auto Dj.

This means if we fix that issue, we can put the xFader disabled or bypass option into the xFader curve calculation, without headache.

The Auto Dj should behave like a real DJ, always ready to let the real DJ take over.
If the real DJ does not use the xFader, the Auto Dj should not use it either.

While I think it is handy to keep a Crossfader Auto Dj mode, in this case the Auto Dj should use the Volume sliders.

Hopefully this full matches the original issue.

@gtrentalancia do you have interest to work for this? An optional volume slider AutoDj mode?
I think it should be straight forward to let the Auto Dj work on an internal xFader and adopt it either to the real xFadet or the volume slider.

Any other ideas?

gtrentalancia commented:
> I have checked this PR and it works as described. However it does not fully resolve the original issue of accidentally touching the hardware xFader.

Hello.

I have tested it again with my external hardware controller and it does fully resolve the original issue of accidentally touching the hardware Crossfader.

When the Crossfader is disabled with the new Mixxx option that I created in this Pull Request, both the software (Mixxx) and hardware (external controller) Crossfader knob becomes ineffective and it does not change the volume of the two decks.

Therefore, the proposed patch fully resolves the issue of accidentally moving the Crossfader.

gtrentalancia commented:
I do not foresee other "urgent" contributions, unless it is a serious bug fix with this change, which seems very unlikely.

In the longer term, I foresee a bug fix related to the Mixxx database and, if you need to move the AutoDJ from Crossfader to Volume Sliders, then I can have a look at it without promising anything, given that I am doing this only as a music hobby in my free time and I am not a professional DJ.

daschuer commented:
Thank you for offering your contributions. 

I understand that this patch entirely fixes the bug for you, since you do not use auto DJ. If you use it, you can still accidentally move the crossfader. 

Compared to the current state, I would prefer the solution to just remove the crossfader mapping from the hardware controller. But let's keep this PR open, maybe one else find it useful. At least it is a good start. 

Would you mind to sign the contributor agreement to allow merging this as part of a whole solution? 
https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ
Please comment here when done. Thank you.

gtrentalancia commented:
But it also depends on how this Pull Request is dealt with, because if the approval process is too sluggish then it might not be worth the time...

gtrentalancia commented:
> Thank you for offering your contributions.
> 
> I understand that this patch entirely fixes the bug for you, since you do not use auto DJ. If you use it, you can still accidentally move the crossfader.

Oh right, now I can understand what you meant !

Sorry for getting you wrong earlier on, I understand your point now...

Yes, of course, given that AutoDJ is still entirely based on the Crossfader this is a limitation of the AutoDJ itself.

However, the human DJ is probably away from the console and controller, therefore it is unlikely that someone moves the knob by accident.

gtrentalancia commented:
In that view, it is still worth investing some time in improving the AutoDJ feature and moving it from using the Crossfader to using the Volume Slider, so at the same time, we get the benefit of not interfering by mistake with "nonlinear" Crossfader settings and other anomalies.

gtrentalancia commented:
> Would you mind to sign the contributor agreement to allow merging this as part of a whole solution?
> https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ
> Please comment here when done. Thank you.

License Agreement Signed.

You can now merge it and when it's done, I will open another PR for the related Manual changes.

gtrentalancia commented:
AutoDJ improvement, maybe next week or ten days, I don't know, but after this is merged anyway...

SQLite database bug, hopefully by the end of the month.

daschuer commented:
> But it also depends on how this Pull Request is dealt with, because if the approval process is too sluggish then it might not be worth the time...

I am very sorry, about rejecting your first contribution like that. It must be a really bad experience. This happens sometimes, because new contributors where not involved in all the discussions we had over the years. In general it is better to propose a new feature idea on https://mixxx.zulipchat.com/ or as a bug at https://bugs.launchpad.net/mixxx first. Than we can discuss the best solution without risk to code for the rubbish bin (You code is still useful btw.)

gtrentalancia commented:
At this point, why not locking the whole controller (and maybe GUI, excluding AutoDJ itself) when AutoDJ is enabled ?

That makes more sense to me...

daschuer commented:
Because the Auto DJ is designed as assistant. You could continue to control anything else while auto DJ is working. It just helps you to load and start the tracks. Unfortunately you currently cannot stop a started transition right now. A pending bug: https://bugs.launchpad.net/mixxx/+bug/995982

All Auto DJ related bus are here: https://bugs.launchpad.net/mixxx/+bugs?field.tag=autodj

ronso0 commented:
> There are enough ways of disabling the crossfader already that we decided an explicit feature was unnecessary. One change that might help with is showing the [L, M, R] crossfader position settings for each channel on two-deck mode. (This is the way users disable the crossfader on hardware mixers, by setting all channels to M or sometimes it's called "thru".) Currently on LateNight, for example, those settings only appear in 4-deck mode.

I agree this would be the easiest way to disable the crossfader, also it's somehow standart due to being implemented in (4-deck) hardware mixers.
Thanks for the reminder, I just picked the work again. Got stuck somehow with the Deere implementation..

Be-ing commented:
Hi @gtrentalancia and thank you for working on Mixxx. However, I agree with @ywwg on this topic. There already are ways to accomplish what this implements and I don't think we should add yet another. One checkbox may not seem like a big deal, but as @ywwg mentioned, every option adds complexity which accumulates over time (both in terms of user experience and code complexity), so we need to consider carefully whenever a new option is proposed. Mixxx's preferences are already quite daunting.

Simply deleting the mapping for the crossfader for your controller is an easy solution, albeit not very obvious to discover. The crossfader assignment switches could be a solution, but their state is not remembered when Mixxx is restarted. Maybe they should be. That would be an easy change to make `orientation` a persistent ControlObject. On that note, the roles of ConfigObjects versus persistent ControlObjects could use further discussion and clarification...

daschuer commented:
@Be-ing: The issue here is that using the deck orientation buttons is not a solution because it breaks auto DJ. It targets also not exactly the issue of accidentally touching the hardware crossfader. Without Changing some Auto DJ code, deleting the crossfader from the mapping would be the best suitable solution, but this requires some insights to do it. 

That's why I think the request for a checkbox solution is reasonable. This is also confirmed by @nschloe here: https://bugs.launchpad.net/mixxx/+bug/1819758.

What could be the best solution? 
In the sens of "Auto DJ is an assistant that behaves like a real DJ" it would be reasonable to add a line fader mode and allow to disable the crossfader via the characteristics curve or the orientation switches. 

Even if we use the orientation switches, it requires thinking around the corner for new users, this is solved by the checkbox proposed here.  I can think of orientation switches presets to target this, a combobox with "Bypass Crossfader" and "Classic Layout" or something. 

This will also avoid to clutter the skins with the orientation switches that are never touched for many users.

daschuer commented:
Thinking about the  combobox idea again I think this will be a natural solution even to replace the hamster style checkbox. We can actually remove it because we have now the deck orientation buttons, which was not the case when introducing the hamster style checkbox. 
The preferences option will allow to not show the orientation buttons in skins and use the space for mor important things: 

The Combobox can be popuated like this, and presets the deck orientation buttons:

Deck Orientation:  
[Regular]
[Hamster]
[Bypass]

In case of Bypass, disable AutoDJ. We may pop up a dialog, if one tries to enabled it. 
In case of bypass, we can remove the xfader GUI slider. 

What do you think? 

Would this be helfpull for the GUI @ronso0?

ronso0 commented:
I'd like to list the aspects that always pop up in my head when we talk about crossfader, and AutoDJ:
**1.** AutoDJ only works well if the user has set a 'mixing' crossfader curve, with 'scratching' curve Auto-fade is just pointless
**2.** user may want an easy way to disable the crossfader, @gtrentalancia pointed out one reason
**3.** do we need an indicator or some protection when users switch to 2 decks, coming from 4-decks mode with unusual deck orientation?

I do like the combobox idea. Preferences > Crossfader would be the natural place to look for that option IMO. We could remove the orientation buttons from 2-deck mode then and pass some responsibility to te users (see **3.**)
However, to make the AutoDJ a pleasant assistant (see **1.**) we should IMO indeed switch to **Vol fader manipulaion** as @daschuer mentioned, instead of using the crossfader. This would save us from changing the crossfadder curve back & forth behind the scenes when users enable AutoDJ, or check for any other crossfader-related settings.
If users spontaneously interfere with AutoDJ (dropping samples or whatever) the crossfader would react as they'd expect, respecting whatever curve they adjusted.
To make this fully work we'd need to **center the crossfader** when AutoDJ is enabled: this would allow reliable mixing no matter what mode is set via the combobox. I think it would help to implement the AutoDJ crossfader highlight from Deere into the other skins as well.
Users that have hidden the mixer, or just the crossfader, probably know what they're doing, so I don't see a need to force-show any components like crossfader or Vol faders apart from the crossfader highlight.

Re **3.**
For me that thought was the reason to implement orientation buttons into LateNight that stay visible in 2-deck mode, and disappeared as soon as they were set to dfault orientation. This not very obvious mechanism lead to some confusion ( ;) @ywwg )
Now, I'm somehow convinced to remove them completely from 2-deck mode.

I'd merge #2075 now as a workaround until the AutoDJ/crossfader changes are inplace. Don't know if I'm of any help with that, but I think that's the way to go to realize a pleasant AutoDJ UX.

ywwg commented:
> Now, I'm somehow convinced to remove them completely from 2-deck mode.

Why do you want to remove them? Note that turntablist DJs are known to put the crossfader in "hamster mode," where the channels are assigned backwards.  We want to make sure we preserve that configuration.

ywwg commented:
> , it requires thinking around the corner for new users,

I'm ok with orientation switches being an optional non-default UI feature, but they serve a useful purpose.

daschuer commented:
>> Now, I'm somehow convinced to remove them completely from 2-deck mode.

> Why do you want to remove them? Note that turntablist DJs are known to put the crossfader in "hamster mode," where the channels are assigned backwards. We want to make sure we preserve that configuration.

It was just the idea to save screen space by moving the controls to the preferences, assuming they are not used during the performance in 2 Deck mode.

The current hamster style solution with a separate bool CO xFaderReverse
is unusual in this sense. The idea was to remove this option and add a helper GUI that configured the orientation switches instead. This makes the xFader configuration more consistent IMHO.

ywwg commented:
OK I think that's a good plan. You're right, this is not an option that I need to change while I'm playing, so moving it out of the skin makes sense.  And once it's a preference, it'll be able to persist.  sounds good to me.

Holzhaus commented:
Is this still relevant? AFAIK it's already possible to disable the crossfader by assigning both decks to "middle".

Be-ing commented:
This was never relevant IMO and we should close it.

lepieru commented:
> AFAIK it's already possible to disable the crossfader by assigning both decks to "middle".

@Holzhaus How do you do this in Mixxx?

Holzhaus commented:
There are lid switches next to the crossfader, at least in LateNight and Deere in Mixxx 2.3 beta (maybe even earlier versions, not sure).
